### PR TITLE
Restock fighter ammo from carrier's cargo hold

### PR DESCRIPTION
**Pull request by tehhowch**
This is an implementation of #1971 with the ammo stockpile in the carrier's cargo - not its installed outfits
As with the installed-outfits implementation (#4128), a fighter that gets resupplied will launch even if it wasn't able to fill up on fuel.

 - Cargo refill allows both carrier and fighter to use the same ammo resource without impacting carrier damage output
    e.g. Carrier (Meteor) & Finch (Meteor)
 - Upon docking, carried ships pool ammo in cargo bay for other fighters to use

(Current) downsides:
 - Mission NPCs cannot be given explicit outfit cargo, thus scripted battles cannot guarantee use of ammo restocking - only that ships in the `fleet` will collectively have said outfits in cargo 
 - Player cargo is filled in reverse size order, beginning with your [first escort][firstescort] (see #2757) and your flagship is filled last and only if necessary
   - Biggest outfits are almost never ammo
   - Carriers would probably be geared to have large cargo holds (so they can hold lots of ammo for fighters)
   - **Ammo would likely end up in ships that fighters do not board without undue micromanagement** of carrier position in the escort list (fighter micromanagement is already a known detraction from their use)
   - No way to tell a fighter to load up on ammo without deploying it (you lose control of your ships' cargo space, e.g. reduced plunder capability)
 - If we resupply fighter ammo from cargo, why not carrier ammo from cargo too?

[firstescort]: https://user-images.githubusercontent.com/20871346/51285272-ed255d00-19b3-11e9-89ed-47df590a3d9f.png

Additional discussions:
https://github.com/zwparchman/endless-sky/pull/1#issuecomment-315591202
https://github.com/endless-sky/endless-sky/issues/1939#issuecomment-269992593
https://github.com/endless-sky/endless-sky/issues/2336
https://steamcommunity.com/app/404410/discussions/2/492378265890026032/
https://steamcommunity.com/app/404410/discussions/0/1319962683445095725/#c1327844097131431281
(I'm sure there are other references as well)

### CODE REVIEW

**Comment by Tadrix** [comment]
I'd prefer them to be resupplied from the cargo, if it's possible to have the carriers automatically stock up on fighter ammo as a cargo assignment priority.

As an alternative, instead of this there could be fighter-specific missile weapons that could be recharged only on carriers, but without consuming anything, for free. As a side effect, this would make fighters more desirable while avoiding inconvenience of having to micromanage them so much.

Another point is,

> Mission NPCs cannot be given explicit cargo

IMO we should have it a long time ago, but as somebody who's not a programmer I have no idea how much work would that involve.

**Comment by petervdmeer** [comment]
I prefer fighters to be re-supplied from a carriers installed outfits; conceptually this also makes sense; cargo holds are just for carrying cargo; you cannot use any items in cargo (scanners, hyperdrives, ammo) during flight. While the ammo-storage-racks are built to allow in-flight reloading of weapons (wherever they are on the carrier, even in fighters).

Gameplay-wise we could change the ammo-storage racks to use up more cargo space and less outfit space if that makes more sense, as is also suggested by Nescio0 in #4128.

**Comment by petervdmeer** [comment]
If we do go the path of reloading launchers from cargo-space, then I feel that such option should apply to all escorts being able to reload ammo from their own and all other escorts cargo spaces.
The boarding mechanism already allows for transfer of ammo between ships and the refueling-request code already contains similar logic.
For this path we would probably also want to introduce some additional settings like the time it takes to move ammo from cargo-space to a launcher. (Think of safely unpacking high-explosive weapons and transferring them safely into a launcher or automated storage rack.)

**Comment by Amazinite** [comment]
I think a more sensible solution as to not make useless the ammo capacity upgrade outfits would be to remove the ability for plundering ammo into your weapon. 

An explanation could be that the guns are in areas that can’t be accessed while in flight (e.g. good luck pulling missiles from your cargo to the wings of an interceptor to reload them). 

Then the reason as to why drones and fighters can reload from cargo is because they can be worked on from the outside when docked; you have free access to the cargo of the carrier and the guns of the carried ships. 

Then we just hand wave away the reloading of externally stored ships, or go full realism and make only internally stored ships be able to reload from a carrier’s cargo. 

This same explanation would mean that reloading from the carrier’s am I storage wouldn’t work, as the storage is probably just as inaccessible as the launcher itself, or else the storage could be refilled from cargo.

File: source/Ship.cpp
Comment (petervdmeer) [inline source/Ship.cpp]: Do we want to re-stock ammo back from fighters to the carriers cargo-space? For a flagship I would consider this undesirable, since the player might want to use the cargo-space for something like plundering instead.

For non-flagships I think it would also be fine to just leave the ammo in the fighter, although I don't have a very strong feeling there either way.

Comment (petervdmeer) [reply source/Ship.cpp]: One other consideration: in the current implementation of planetary cargo-loading there is no guarantee that ammo returned to the carriers cargo-hold will again be loaded into the carrier on launch.

**Comment by petervdmeer** [comment]
@Amazinite Yes, we can come up with good in-game explanations for either option:
- Current situation (no re-loading): Outfitted ammo is just different/unique for every ship.
- From cargo (fighters only): Your explanation.
- From cargo (fighters and ship itself): Additional keyword (on launchers and/or storage) like `reload from cargo time` as indicated by me above.
- From cargo (any ship to any ship): Ammo-transfers similar to fuel transfers (and `reload from cargo time` keyword).
- From outfits (fighters and ship itself): Ammo loading in-flight can only be done from storage racks designed for it.
- From outfits (also in-between ships): Ammo transfers in-flight is only safe for escorts when done from storage racks.

My preferences is **From outfits (fighters and ship itself)** since it is very predictable, or **From cargo (any ship to any ship)** since it is also very predictable, ammo stored in cargo anywhere can be loaded anywhere.

**Review by Amazinite — APPROVED** [review]
Did a poll on Discord about this and its sister PR and the results came out to people wanting fighters to rearm from cargo. I've gone back and forth on where to rearm from, hence why I threw the poll up, and I think I've ultimately settled on rearming from cargo as well.

**Comment by Pointedstick** [comment]
I think this makes more sense too, personally. Imagining weapons techs pulling missiles out of their launch tubes mid-battle to re-arm fighters fighter strikes me as something they would demand extra hazard pay for. :)

**Comment by EjoThims** [comment]
Would it be possible to make fighters that have run out of ammo automatically dock with nearest/any carrier carrying their needed ammo?

Or would that be a separate issue?

**Comment by daveeater** [comment]
Very late to the discussion here, but it seems to me that fighters should re-load from carrier magazines, not from cargo.  "Magazines" are effectively storage racks.  But no big deal - if you want to stock your carrier's cargo space with ammo, you purchase outfit expansions and then purchase (and stock) storage racks.  Fighters should be able to re-arm from carrier storage racks (magazines) - that's a legitimate function of a hangar deck/space.

**Comment by quyykk** [comment]
What's the status on this? It would be nice to have this in the game. Although, I prefer the other PR. I don't see an issue with fighters taking ammo from the carriers stock since that's the role of the carrier. It is also not consistent with how normal ships use ammo. One of the downsides listed on the other PR is:

> Requires either ammo storage outfits or the carrier to have the same weapon as the fighter

But I fail to see how this is an issue. It is only logical for a carrier to have ammo storage outfits for its fighters.

**Comment by Amazinite** [comment]
About a month after my approval of this PR I thought, "you know what, why not just allow both options?" So I have a [branch](https://github.com/Amazinite/endless-sky/tree/rearm-preference) that would add a preference for selecting the type of rearming you want (cargo only, installed only, cargo first, installed first, or no rearming), although that preference doesn't actually do anything on that branch. So to get everything together, someone needs to merge together that branch, this PR, and #4128.

**Comment by EjoThims** [comment]
>cargo first

Of all the options, this would be my preference, by far.

But would it not be possible (and faster and easier, too) to go ahead and finally merge this, and then get the other options and the selection bundled together to merge later?

**Comment by quyykk** [comment]
Restocking from cargo has tons of disadvantages. Not only from a consistency standpoint but also because of issues like this

> Ammo would likely end up in ships that fighters do not board without undue micromanagement

There is currently no way to control where cargo ends up in without abusing various mechanics. This would make player owned carriers almost impossible and very painful to use as part of a player fleet. See also all the other disadvantages listed on this PR.

Like honestly, this PR is a serious game design flaw imo. I simply do not understand why this PR is more popular than the other one or why it is being considered. Can someone explain it to me please?

**Comment by Hurleveur** [comment]
> Restocking from cargo has tons of disadvantages. Not only from a consistency standpoint but also because of issues like this
> 
> > Ammo would likely end up in ships that fighters do not board without undue micromanagement
> 
> There is currently no way to control where cargo ends up in without abusing various mechanics. This would make player owned carriers almost impossible and very painful to use as part of a player fleet. See also all the other disadvantages listed on this PR.
> 
> Like honestly, this PR is a serious game design flaw imo. I simply do not understand why this PR is more popular than the other one or why it is being considered. Can someone explain it to me please?

Well just put carriers at the end of your fleet and don't be cargo running at the same time.
If it was only from outfits, this mechanic would be almost pointless according to me, don't want to waste precious outfit space for missiles.
In the worst case, can't we change how cargo works/add an outfit that allows outfitting missiles by taking cargo space?
The later

**Comment by petervdmeer** [comment]
This PR and its counterpart #4128 have been open for more than 3 years now. We have had good discussions in many places (GitHub, Steam, Discord, etc) providing arguments in favor and against either way ("from outfits" or "from cargo") of reloading fighters from carriers.

I don't expect many new arguments to show up in the near future and with the recent addition of new code-maintainers I do see an opportunity to get "reloading fighters from carriers" merged soon.
The most desired option seems to be "from cargo" (as implemented in this PR), but most discussions also show it as currently not-workable, because it is very hard to control which cargo gets stored in which ship. That leaves the "from outfits" option as the only short-term merge-able solution.

We can always revise in the long-term when the context changes, but I'm closing this PR now in favor of trying to merge #4128 soon.

**Pull request closed without merging**
